### DISCUSSION TITLE
Rendering Subcontrols

### DISCUSSION THREAD

**Discussion by chris20242024** [FactoryXCode/MfPack#65] [General]
Hello,
I`m searching for a way to control the Rendering  Subcontrols of LineIn/Microphone (Volume/Mute)
I know its possible on the Recording Devices . But not on the Rendering Devices.(Rendering will only Show the Main Devices and not the SubControls (as on the Recording Side). How is that possible?
![Rendering-Mixer](https://github.com/FactoryXCode/MfPack/assets/170936985/e6680ab9-a1dc-4183-badc-a4d8e38b7ea3)

Best Regards Chris

**Comment by FactoryXCode** [comment]
Hi. 
You may have a look at the MfpMixer component sample (MfpMixer.pas).
Also there is an audio device topo utils unit (WinApi.CoreAudioApi.DevTopoUtils.pas),
that contains the most used methods to gather the audio endpoints and their properties.
All hardware devices - rendering (eRender), capture (eCapture) or both (eAll) - are called endpoints, 
where each endpoint may have pins where for example a mute control is connected with.
For this you might use the IKsJackDescription2 interface see [here](https://learn.microsoft.com/en-us/windows/win32/api/devicetopology/nn-devicetopology-iksjackdescription2)
So, you have to iterate through all of these, to get those sub-controls and their state. 

In short terms:
The most logical way is, to search for all capture devices followed by rendering devices.
And when you got those then iterate for their properties, pins and state.

Regards Tony.

**Reply by chris20242024** [reply]
Hi, many thanks for the answer.
Where can I find the MfpMixer component sample (MfpMixer.pas)?.
I can`t find it in the MfPack/Samples Directory.
Regards Chris

**Reply by FactoryXCode** [reply]
The MfPMixer is located [here](https://github.com/FactoryXCode/MfPack/tree/Master/MfPack/Samples/MfComponents)

Note: Also read about the [devicetopology (WinApi.CoreAudioApi.DeviceTopology.pas) ](https://learn.microsoft.com/en-us/windows/win32/api/devicetopology/)  how to set for example mute, loudness etc. if the hardware support those.
